### PR TITLE
Add new config mechanism and use for S3 scratch bucket

### PR DESCRIPTION
**Pull request by vinceatbluelabs**
This is more generic configuration mechanism for records-mover that can be set system-, user-, or session-wide as needed.  It'd replace the current `/usr/local/bin/scratch-s3-url` mechanism.

### CODE REVIEW

File: records_mover/session.py [hunk 1]
```diff
@@ -41,6 +42,33 @@ def _infer_scratch_s3_url(session_type: str) -> Optional[str]:
     if "SCRATCH_S3_URL" in os.environ:
         return os.environ["SCRATCH_S3_URL"]
 
+    # TODO: do I want this in 'app.ini'?
```
Comment (bluelabsbutler) [inline records_mover/session.py:45]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/session.py [hunk 2]
```diff
@@ -41,6 +42,33 @@ def _infer_scratch_s3_url(session_type: str) -> Optional[str]:
...
+    # TODO: where else to document this?
```
Comment (bluelabsbutler) [inline records_mover/session.py:46]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/session.py (hunk 1 above)
Comment (vinceatbluelabs) [reply records_mover/session.py:45]: Seems a little generic, but this seems to be config_resolver's default, and I guess having a directory named records_mover leaves room for additional files inside, so not sure I have a better idea!

File: records_mover/session.py (hunk 2 above)
Comment (vinceatbluelabs) [reply records_mover/session.py:46]: I'm kind of thinking we create a CONFIG.md describing (for now) how to configure the S3 bucket, including the content below, and point to it on github.com in the error message when we raise a NoTemporaryBucketConfiguration or CredsDoNotSupportS3Export.  Seem reasonable?

**Comment by vinceatbluelabs** [comment]
Hey @cwegrzyn, could you take a look and provide some feedback on this so far?  

I haven't yet looked into the db-facts aspect of this, but it seems to capture the essence of what we talked about for configuration.

**Comment by vinceatbluelabs** [comment]
Totally!  Given AWS takes credentials alongside the S3 bucket, (except maybe in GovCloud?), I could certainly see reasons for folks to want to have that level of configurability.  One flag is that AWS credentials would often need to change along with that S3 bucket configuration - having a consistent way to address this would be great, too, instead of relying on folks handing records-mover AWS creds that match the database they pointed it to.

Of course, sometimes this problem is thorny on its own - e.g., with third party tools to provide temporary AWS credentials.  This sounds a little more in the realm of what something like db-facts would provide with the ability to call out to scripts to fetch back creds and all.

File: docs/CONFIG.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+### Configuring db-facts
+
+[db-facts](https://github.com/bluelabsio/db-facts) is a complementary
+project used to configure database credentials.  Please see
+[db-facts documentation](https://github.com/bluelabsio/db-facts/blob/master/CONFIGURATION.md)
+for details on configuration.
+
+## Temporary locations
+
+TODO: Finish the rest here
```
Comment (bluelabsbutler) [inline docs/CONFIG.md:67]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: docs/CONFIG.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+TODO: Add `SCRATCH_S3_URL` env variable
```
Comment (bluelabsbutler) [inline docs/CONFIG.md:69]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/session.py [hunk 5]
```diff
@@ -36,11 +37,32 @@ def _infer_session_type() -> str:
...
-    scratch_s3_url: Optional[str]
-
     if "SCRATCH_S3_URL" in os.environ:
         return os.environ["SCRATCH_S3_URL"]
 
+    # config_resolver logs at the WARNING level for each time it
+    # attempts to load a config file and doesn't find it - which given
+    # it searches a variety of places, is quite noisy:
+    #
+    #   group=bluelabs:app=records_mover:Skipping unreadable file /etc/bluelabs/records_mover/app.ini (File not found)
```
Comment (bluelabsbutler) [inline records_mover/session.py:47]: E501 line too long (118 > 100 characters)

File: records_mover/session.py [hunk 6]
```diff
@@ -36,11 +37,32 @@ def _infer_session_type() -> str:
...
+    #   group=bluelabs:app=records_mover:Skipping unreadable file /etc/xdg/bluelabs/records_mover/app.ini (File not found)
```
Comment (bluelabsbutler) [inline records_mover/session.py:48]: E501 line too long (122 > 100 characters)

File: records_mover/session.py [hunk 7]
```diff
@@ -36,11 +37,32 @@ def _infer_session_type() -> str:
...
+    #   group=bluelabs:app=records_mover:Skipping unreadable file /Users/broz/.config/bluelabs/records_mover/app.ini (File not found)
```
Comment (bluelabsbutler) [inline records_mover/session.py:49]: E501 line too long (133 > 100 characters)

File: tests/unit/test_session.py [hunk 8]
```diff
@@ -2,15 +2,16 @@
 from records_mover import Session
 import unittest
 
-
+@patch('records_mover.session.get_config')
```
Comment (bluelabsbutler) [inline tests/unit/test_session.py:5]: E302 expected 2 blank lines, found 1

File: metrics/bigfiles_high_water_mark [hunk 9]
```diff
@@ -1 +1 @@
-968
+969
```
Comment (vinceatbluelabs) [inline metrics/bigfiles_high_water_mark:1]: One extra line in setup.py for the new dependency

File: records_mover/session.py [hunk 10]
```diff
@@ -36,11 +37,26 @@ def _infer_session_type() -> str:
...
+    cfg = config_result.config
+    if 'aws' in cfg:
+        s3_scratch_url: Optional[str] = cfg['aws'].get('s3_scratch_url')
+        if s3_scratch_url is not None:
+            return s3_scratch_url
+    else:
+        logger.debug('No config ini file found')
+
     if session_type == 'cli':
         try:
```
Comment (vinceatbluelabs) [inline records_mover/session.py:61]: There will be a second PR to retire this feature once we have internal uses switched over to the config file.

File: records_mover/session.py [hunk 11]
```diff
@@ -36,11 +37,26 @@ def _infer_session_type() -> str:
...
     if "SCRATCH_S3_URL" in os.environ:
         return os.environ["SCRATCH_S3_URL"]
 
+    # config_resolver logs at the WARNING level for each time it
+    # attempts to load a config file and doesn't find it - which given
+    # it searches a variety of places, is quite noisy.
+    #
+    # https://github.com/exhuma/config_resolver/blob/master/doc/intro.rst#logging
+    #
+    # https://github.com/exhuma/config_resolver/issues/69
```
Comment (vinceatbluelabs) [inline records_mover/session.py:49]: I have a PR open - hopefully it gets accepted: https://github.com/exhuma/config_resolver/pull/70

**Comment by vinceatbluelabs** [comment]
OK!  Updated and ready to merge.

File: records_mover/session.py (hunk 11 above)
Comment (exhuma) [reply records_mover/session.py:49]: Should be fine now. Let me know if there are any remaining issues.

**Pull request merged**